### DISCUSSION TITLE
Is it possible to extract embeddings?

### DISCUSSION THREAD

**Discussion by abtawfik** [LukasZahradnik/PyNeuraLogic#33] [Q&A] [answered]
I am playing around with this library but am curious if it is possible to extract node level or graph-level embeddings? I looked at the methods attached to the `evaluator` but the best I could find was `evaluator.neuralogic_model.neural_model.getAllWeights().toString()` which returned a string of the weights for each level.

The goal here is to perform some clustering on the examples. I very much appreciate any response and would like to expand using his library a lot more and potentially present it to my team once I can wrap my head around it.

**Comment by LukasZahradnik** [comment, marked as answer]
Hello. Yes, it should be possible to extract all learnable parameters using the `state_dict` method on your model or evaluator (`evaluator.state_dict()`).

I realized it might be better to give it a more apparent name, so in the new version (0.1.7) that I just released, there is an alias `parameters` for `state_dict` (`evaluator.parameters()`) as well as fixed retrieving parameter values of one specific type - might be good to update your version.

With the `state_dict`/`parameters` method, you will get a dictionary with keys `"weights"` (which contains actual parameters' values) and `"weight_names"`.

**Reply by abtawfik** [reply]
Well that was simple!
Thank you and I'd be remiss if I didn't also say how much I enjoyed the elegance of the lifted relational neural network framework.

**Reply by LukasZahradnik** [reply]
Glad you enjoyed it!
